modules/xfd.js

```diff
@@ -0,0 +1,2052 @@
+            var text;
+            // add date header if the log is found to be empty (a bot should do this automatically)
+            if (!pageobj.exists()) {
+                text = '{{subst:AfD log}}\n' + added_data;
+            }
+            else {
+                var old_text = pageobj.getPageText() + '\n'; // MW strips trailing blanks, but we like them, so we add a fake one
+                text = old_text.replace(/(<!-- Add new entries to the TOP of the following list -->\n+)/, '$1' + added_data);
+                if (text === old_text) {
+                    var linknode = document.createElement('a');
+                    linknode.setAttribute('href', mw.util.getUrl('Wikipedia:Twinkle/Fixing AFD') + '?action=purge');
+                    linknode.appendChild(document.createTextNode('How to fix AFD'));
+                    statelem.error(['Could not find the target spot for the discussion. To fix this problem, please see ', linknode, '.']);
+                    return def.reject();
+                }
+            }
+            pageobj.setPageText(text);
+            pageobj.setEditSummary('Adding [[:' + params.discussionpage + ']].');
+            pageobj.setChangeTags(Twinkle.changeTags);
+            pageobj.setWatchlist(Twinkle.getPref('xfdWatchList'));
+            pageobj.setCreateOption('recreate');
+            pageobj.save(def.resolve, def.reject);
+        });
+        return def;
+    };
+    Afd.prototype.addToDelsortLists = function () {
+        var params = this.params;
+        var defs = new Array(params.delsortCats.length);
+        for (var i = 0; i < defs.length; i++) {
+            // ugly, would be a lot better when mb.w.page#load and #save return promises
+            defs[i] = $.Deferred();
+        }
+        params.delsortCats.forEach(function (cat, idx) {
+            var delsortPage = new Morebits.wiki.page('Wikipedia:WikiProject Deletion sorting/' + cat, 'Adding to list of ' + cat + '-related deletion discussions');
+            delsortPage.setFollowRedirect(true); // In case a category gets renamed
+            delsortPage.load(function (pageobj) {
+                var discussionPage = params.discussionpage;
+                var text = pageobj.getPageText().replace('directly below this line -->', 'directly below this line -->\n{{' + discussionPage + '}}');
+                pageobj.setPageText(text);
+                pageobj.setEditSummary('Listing [[:' + discussionPage + ']].');
+                pageobj.setChangeTags(Twinkle.changeTags);
+                pageobj.setCreateOption('nocreate');
+                pageobj.save(defs[idx].resolve, defs[idx].resolve); // failures aren't important
+            });
+        });
+        return $.when.apply($, defs);
+    };
+    Afd.prototype.patrolPage = function () {
+        if (Twinkle.getPref('markXfdPagesAsPatrolled')) {
+            new Morebits.wiki.page(Morebits.pageNameNorm).triage();
+        }
+        return $.Deferred().resolve(); // XXX
```

Comment:
should this not wait until after patrolling is done? (If this is a carry-over from the prior behavior, feel free to ignore)

Reply:
Yes, but Morebits.wiki.page#triage doesn't support callbacks or promises (till now), so there's no way to wait.